classes/identity/Identity.inc.php

```diff
@@ -18,112 +18,166 @@
-		} else {
-			return ($salutation != ''?"$salutation ":'') . "$firstName " . ($middleName != ''?"$middleName ":'') . $lastName . ($suffix != ''?", $suffix":'');
+	function &getLocalizedData($key, $preferredLocale = null) {
+		if (is_null($preferredLocale)) $preferredLocale = AppLocale::getLocale();
+		$localePrecedence = array($preferredLocale);
+		// the users register for the site, thus
+		// the site primary locale is the default locale
+		$siteDao = DAORegistry::getDAO('SiteDAO');
+		$site = $siteDao->getSite();
+		if (!in_array($site->getPrimaryLocale(), $localePrecedence)) $localePrecedence[] = $site->getPrimaryLocale();
+		// for settings other than givenName, familyName and affiliation (that are required for registration)
+		// consider also the context primary locale
+		if (!in_array(AppLocale::getPrimaryLocale(), $localePrecedence)) $localePrecedence[] = AppLocale::getPrimaryLocale();
+		foreach ($localePrecedence as $locale) {
+			if (empty($locale)) continue;
+			$value =& $this->getData($key, $locale);
+			if (!empty($value)) return $value;
+			unset($value);
 		}
-	}
 
+		// Fallback: Get the first available piece of data.
+		$data =& $this->getData($key, null);
+		foreach ((array) $data as $dataValue) {
+			if (!empty($dataValue)) return $dataValue;
+		}
 
-	/**
-	 * Get first name.
-	 * @return string
-	 */
-	function getFirstName() {
-		return $this->getData('firstName');
+		// No data available; return null.
+		unset($data);
+		$data = null;
+		return $data;
 	}
 
 	/**
-	 * Set first name.
-	 * @param $firstName string
+	 * Get the identity's localized complete name.
+	 * Includes given name and family name.
+	 * @param $familyFirst boolean False / default: Givenname Familyname
+	 * 	If true: Familyname, Givenname
+	 * @param $defaultLocale string
+	 * @param $prefered boolean If the prefered public name should be used, if exist
+	 * @return string
 	 */
-	function setFirstName($firstName) {
-		$this->setData('firstName', $firstName);
+	function getFullName($familyFirst = false, $defaultLocale = null, $prefered = true) {
+		$locale = AppLocale::getLocale();
+		if ($prefered) {
+			$preferedPublicName = $this->getPreferedPublicName($locale);
+			if (!empty($preferedPublicName)) return $preferedPublicName;
+		}
+		$givenName = $this->getGivenName($locale);
+		if (empty($givenName)) {
+			if (is_null($defaultLocale)) {
+				// the users register for the site, thus
+				// the site primary locale is the default locale
+				$siteDao = DAORegistry::getDAO('SiteDAO');
+				$site = $siteDao->getSite();
+				$defaultLocale = $site->getPrimaryLocale();
+			}
+			$locale = $defaultLocale;
+			$givenName = $this->getGivenName($locale);
+		}
+		$familyName = $this->getFamilyName($locale);
+		if ($familyFirst) {
+			return ($familyName != ''?"$familyName, " :'') . $givenName;
+		} else {
+			return $givenName . ($familyName != ''?" $familyName" :'');
+		}
```

Comment:
Hmmm... I am not sure about this because we use this function also for some third party services that would expect the names to be like that i.e. are somehow based on that western standard. For example ONIX4DOI standard used in mEDRA supports both PersonName and InvertedPersonName (that is then above all used for sorting). Or, for example, in OAI_DC it is good to use "familyName, givenName" in order to be compatible with others. Thus it feels like it would be better that we have the control over it.
Also, it could be that there is no family name, so in that case there actually should be no "," in the locale key, but we would have to check it and provide it with the variable $familyName.
Hmmm... :thinking:

Reply:
Ah ok, best to leave it alone then. :+1: